### PR TITLE
Sanitize bash for volume restore in Dockerfiles

### PR DESCRIPTION
**Pull request by stlaz**
----
This should help us test Travis on multiple Dockerfile changes

### CODE REVIEW

**Comment by stlaz** [comment]
Fails because we can't access RHEL repos.

**Comment by adelton** [comment]
What sort of problems does the current formatting pose to Travis?

File: init-data [hunk 1]
```diff
@@ -193,6 +193,11 @@ if ! [ -f /etc/ipa/ca.crt ] ; then
 		echo "--debug" >> $OPTIONS_FILE
 	fi
 fi
+
```
Comment (adelton) [inline init-data]: This should likely go to separate pull request.

**Comment by adelton** [comment]
Needinfo for 15 days, closing. ;-)

**Comment by stlaz** [comment]
I'll address this hopefully on the morrow.

**Comment by stlaz** [comment]
@adelton the formatting does not pose a problem, I only had this change in a drawer and thought it might be useful to originally test Travis set up with it.
The rationale behind the patch is really just to make the code a bit more readable to anyone trying to improve/understand the Dockerfile.

**Comment by adelton** [comment]
Aha.
I might actually prefer to put this into some utility script and `COPY` + `RUN` that. We will be making the `Dockerfile.*`s longer in the future with the read-only changes, and this seems to be well-contained to justify its own utility.

**Comment by stlaz** [comment]
Yes, I was actually originally thinking about something similar. It'd be great if we could actually source a script for a Dockerfile and then just use the script's functions in the Dockerfile, but maybe we can just create a directory containing utility scripts instead.

**Comment by stlaz** [comment]
That being said - yes, I'll create a util script for this.

**Comment by adelton** [comment]
I'm already creating `tests/` in my other work for CI scripts so having `util/` or something similar would be good.

**Comment by stlaz** [comment]
Not sure whether this is a common practice, but what do you think about the solution where the more complex functions are imported from a bash source file to be called in the Dockerfile (=> current state of this PR being a PoC of that approach).
IMO it's easier to get rid of this one file during build instead of having to keep track of multiple utility scripts from a folder. I also do not suspect that there would be multitude of such utility functions to exceed a reasonable amount of line for the source file, or would it?

**Comment by adelton** [comment]
The removal of the file after it was used does not help saving space because it is already stored in the previous layer(s) when it's `COPY`ed into the image. It will only add confusion because the file that tweaked the image is now gone.
I also don't like `source`ing the file and calling a funtion from it -- I'd prefer plain utility script, perhaps with some parameters if you need it to do different actions at different stages.

**Comment by adelton** [comment]
`utils/` and `/usr/local/bin` would likely be the best.
It'd be nice if the scripts (or script, with some parameters) could be used to both move things aside in build-time and then be used to populate (and re-populate) `/data` in run-time.
Also, in https://github.com/freeipa/freeipa-container/pull/228 we will eventually do the same thing we do for `/data` for things that don't need to persist, using `volume-tmp-list`. So parametrizing the script to be usable for both `/data` and `/tmp` would be great.

**Comment by adelton** [comment]
FYI, the code that I'm currently working with is

```
#!/bin/bash

set -e

LIST="$1"
VOLUME="$2"

if [ -z "$LIST" -o -z "$VOLUME" ] ; then
	echo "Usage: $0 volume-list-file /volume-path" >&2
	exit 1
fi

mkdir -p "$VOLUME-template"
while read i ; do
	echo "$i"
	parent=$( dirname "$i" )
	if [ -e "$i" ] ; then
		tar cf - ".$i" | ( cd "$VOLUME-template" && tar xvf - )
	else
		mkdir -p "$VOLUME-template"/"$parent"
	fi
	mkdir -p "$parent"
	rm -rf "$i"
	ln -sv "$VOLUME${i%/}" "${i%/}"
done < "$LIST"
```

But I need to figure out what's the deal with `/var/log/journal` -- I know that in the past `docker build` (?) was failing when we made `/var/log` into a symlink to `/data/var/log` which did not exist (in build time) and `oci-systemd-hook` wanted to mount things there.

But I don't seem to get that error today and at the same time I'm not certain where we persist the journal from the container at all. That will likely need to be investigated first.

**Comment by adelton** [comment]
The master branch in 8a250f9190be89b00852e281fc16a37b38201493 introduced `utils/prepare-volume-template` and `utils/populate-volume-from-template`.

**Pull request closed without merging**